core/src/main/java/org/infinispan/CacheStream.java

```diff
@@ -0,0 +1,251 @@
+      /**
+       * Method invoked when the segment has been found to be consumed properly by the terminal operation.
+       * @param segment The segments that were completed
+       */
+      void segmentCompleted(Set<Integer> segment);
+   }
+
+   /**
+    * {@inheritDoc}
+    * <p>This operation is performed remotely on the node that is the primary owner for the key tied to the entry(s)
+    * in this stream.</p>
+    * <p>NOTE: This method while being rehash aware has the lowest consistency of all of the operators.  This
+    * operation will be performed on every entry at least once in the cluster, as long as the originator doesn't go
+    * down while it is being performed.  This is due to how the distributed action is performed.  Essentially the
+    * {@link CacheStream#distributedBatchSize} value controls how many elements are processed per node at a time
+    * when rehash is enabled. After those are complete the keys are sent to the originator to confirm that those were
+    * processed.  If that node goes down during/before the response those keys will be processed a second time.</p>
+    * <p>This method is ran distributed by default with a distributed backing cache.  However if you wish for this
+    * operation to run locally you can use the {@link CacheStream#iterator()} method to return all of the results
+    * locally and then use {@link Iterator#forEachRemaining(Consumer)} method for a single threaded variant.  If you
+    * wish to have a parallel variant you can use {@link java.util.stream.StreamSupport#stream(Spliterator, boolean)}
+    * passing in the spliterator from the stream.  In either case remember you <b>must</b> close the stream after
+    * you are done processing the iterator or spliterator..</p>
+    * @param action
+    */
+   @Override
+   void forEach(Consumer<? super R> action);
+
+   /**
+    * {@inheritDoc}
+    * <p>Usage of this operator requires closing this stream after you are done with the iterator.  The preferred
+    * usage is to use a try with resource block on the stream.</p>
+    * <p>This method has special usage with the {@link org.infinispan.CacheStream.SegmentCompletionListener} in
+    * that as entries are retrieved from the next method it will complete segments.</p>
+    * <p>This method obeys the {@link CacheStream#distributedBatchSize(int)} setting by only ever returning the
+    * elements that mapped to that many keys.  Note that when using methods such as
+    * {@link CacheStream#flatMap(Function)} that you will have possibly more than 1 element mapped to a given key
+    * so this doesn't guarantee that many number of entries are returned per batch.</p>
+    * @return the element iterator for this stream
+    */
+   @Override
+   Iterator<R> iterator();
+
+   /**
+    * {@inheritDoc}
+    * <p>Usage of this operator requires closing this stream after you are done with the spliterator.  The preferred
+    * usage is to use a try with resource block on the stream.</p>
```

Comment:
Hmmm, I just saw the comment in the `iterator()` javadoc and I started wondering whether the segment completion listener is notified for spliterator.tryAdvance(). I see from your answer that the iterator note is more specific and the only difference is that the iterator delays the notification until (just before?) `next()` returns the last value in the segment.